packages/std/src/encoding.rs

```diff
@@ -17,9 +16,11 @@ impl Binary {
     /// take an (untrusted) string and decode it into bytes.
     /// fails if it is not valid base64
     pub fn from_base64(encoded: &str) -> StdResult<Self> {
-        let binary = base64::decode(&encoded).context(Base64Err {})?;
+        let binary =
+            base64::decode(&encoded).map_err(|e| InvalidBase64 { msg: e.to_string() }.build())?;
```

Comment:
This I find not-very ergonomic (just searched for this example as a place where source was turned into a string - the point is more meaningful for places where the error is used often, like Parse?).

A simple solution would be to export the transform itself from `errors.rs`:

```rust
pub fn invalid_base_64(source: base64::DecodeError) -> StdError {
    InvalidBase64{ msg: source.to_string() }.build()
}
```

not so much code there, but then gives us a nice alternative to `context`, I think they are about the same:

```rust
let binary = base64::decode(&encoded).context(Base64Err {})?;

let binary = base64::decode(&encoded).map_err(invalid_base_64)?;
```

What do you think? Or do you have a better idea?

Reply:
There are multiple aspects to this and I don't have a great solution yet:
- Right now, we expose some error constructor structs like Base64Err. For other errors we only expose maker functions like `unauthorized()`. This is pretty inconsistent.
- I'd be happy if we can avoid exposing the snafu-specific magically generated error structs (to reduce dependency on snafu, especially in the public interface)
- With the `to_vec`/`from_slice` and similar helpers in place, there are actually very few usages of `.context(` in the whole codebase
- The helpers we currenty have return Results, but `map_err` needs an error instance. So we either need two functions for every error, or always return the error and then use like `return Err(unathorized());` (which would make much sense)
- If the helpers take the original error type, we keep the dependency on one particular implementation in the API. To avoid that, we at least need:

```rust
pub fn invalid_base_64(source: String) -> StdError {
    InvalidBase64{ msg: source }.build()
}

// ...

let binary = base64::decode(&encoded).map_err(|e| invalid_base_64(e.to_string()))?;
```

Reply:
By the way, the `.build()` helper [is very fresh (snafu 0.6.6)](https://github.com/shepmaster/snafu/blob/master/CHANGELOG.md#066---2020-04-05) and was not yet available when the `.fail()` helpers where created

Reply:
> Right now, we expose some error constructor structs like Base64Err. For other errors we only expose maker functions like unauthorized(). This is pretty inconsistent.
> I'd be happy if we can avoid exposing the snafu-specific magically generated error structs (to reduce dependency on snafu, especially in the public interface)

I agree completely, we can just expose function and keep the auto-generated snafu struct private.

Reply:
> If the helpers take the original error type, we keep the dependency on one particular implementation in the API. 

Good point. Then what about taking `Display` as the type, which implies the `to_string` method. Or, even safe, require [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html), which includes Display but ensures we are passing in some error.

Then we have:

```rust
pub fn invalid_base_64(source: std::error::Error) -> StdError {
    InvalidBase64{ msg: source.to_string() }.build()
}

// ...

let binary = base64::decode(&encoded).map_err(invalid_base_64)?;
```

Clean public API, no dependencies on a particular source error.

Reply:
Done in https://github.com/CosmWasm/cosmwasm/pull/305/commits/356b6ff6428466debf9bb34501ba990b1f8a09e4. `Display` turns out to be handy, as this covers both `Error`, `String` and `&str` ([pub trait Error: Debug + Display](https://doc.rust-lang.org/src/std/error.rs.html#47))